fix(typecheck): improve error message when tsc outputs help text

Ujjwaljain16 opened the pull request:
When tsconfig.json is missing and tsc cannot find any configuration file (including in parent directories), it outputs its help text instead of type checking errors. This was being captured and displayed as a confusing 'Typecheck Error' with the entire help text dump (100+ lines of tsc usage instructions).

Changes:
1. Added help text detection in prepareResults() method
   - Checks for 'The TypeScript Compiler - Version' header
   - Checks for 'COMMON COMMANDS' section
   - Both are present in tsc help output

2. Throws clear, actionable error message with:
   - Explanation of what went wrong
   - The configured tsconfig path (or default 'tsconfig.json')
   - Three possible solutions: a. Ensure tsconfig exists in project root b. Verify custom tsconfig path in vitest config c. Check tsconfig is valid JSON

3. Added stderr capture in spawn() method
   - Previously only stdout was captured
   - Configuration errors from tsc may appear on stderr
   - Ensures all tsc output is examined for help text

Testing:
- All existing tests pass (7/7 in test/typescript)
- Manually verified tsc outputs help text when tsconfig missing
- Automated test not included due to tsc's parent directory search behavior making it impractical to reliably trigger this scenario

Note: This is different from TS5058 error which occurs when a specific tsconfig path is provided but doesn't exist. The help text scenario happens when tsc can't locate any config file at all.

### Description

<!-- Please insert your description here and provide especially info about the "what" this PR is solving -->

Resolves #8981

<!-- You can also add additional context here -->

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.
- [ ] Please check [Allow edits by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to make review process faster. Note that this option is not available for repositories that are owned by Github organizations.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

sheremet-va reviewed (changes requested):
Please, add a test that confirms the behaviour is now correct.

Ujjwaljain16 commented:
The failing test "boolean flag 100 should not crash CLI" is unrelated to the typecheck changes in this PR. This PR only modifies typecheck/typechecker.ts for error handling, while the failure is in coverage threshold reporting. The test is expecting "ERROR: Coverage for functions" but it's not appearing in stderr.

All typecheck-related tests pass successfully, including the previously failing Windows CI test

Ujjwaljain16 commented on test/typescript/test/typechecker.test.ts:
```diff
@@ -0,0 +1,95 @@
+import type { TestProject } from '../../../packages/vitest/src/node/project'
```
The “help text” scenario cannot be integration-tested because `tsc` always
searches parent directories for tsconfig.json files. Even with runInlineTests,
tsc still inherits a real working directory and climbs to the Vitest monorepo
root, where tsconfig files exist.

Because of this, tsc will never emit its help text in CI or in Vitest’s test
environment, and automated tests cannot reliably trigger this condition.
@sheremet-va

Ujjwaljain16 replied:
@sheremet-va Thank you for the suggestion! I attempted the invalid tsconfig approach but ran into a technical constraint I'd like your input on.

## Problem: tsc Help Text Trigger Conditions

After testing, I found that `tsc` only outputs help text under these conditions:
- Invoked with zero arguments: `tsc`
- Invoked with unrecognized flags: `tsc --unknownFlag`
- Invoked with `--help`

However, Vitest's typechecker always invokes tsc with valid arguments:
```typescript
// From typechecker.ts spawn()
const args = [
  '--noEmit',
  '--pretty', 'false',
  '--incremental',
  '--tsBuildInfoFile', '...',
  '-p', resolve(root, typecheck.tsconfig)
]
const child = x(typecheck.checker, args, {...})
```

## Invalid `tsconfig.json` Attempts

| Approach | `runInlineTests` Config | `tsc` Output | Result |
|--------|-------------------------|--------------|--------|
| **Syntactically invalid JSON** | `'tsconfig.json': '{ invalid }'` | N/A — Vite/esbuild fails first | ❌ Build error before `tsc` runs |
| **Empty object** | `'tsconfig.json': '{}'` | `error TS18003: No inputs were found` | ❌ Error output, not help text |
| **Missing `compilerOptions`** | `'tsconfig.json': '{"files":[]}'` | `error TS18003: No inputs were found` | ❌ Error output, not help text |

### Root Cause

When `tsc` is invoked with `-p <file>` and the config file is **structurally valid JSON** (even if semantically incorrect), it enters normal compiler mode. In this mode, `tsc` emits **specific diagnostic errors** and never falls back to printing help text.

Help text is only produced when `tsc` itself is invoked with **invalid or missing CLI arguments**, which does not occur in Vitest’s invocation path.


Current Solution: Test Executable Stub
I implemented a temporary executable that outputs tsc help text:
```
// Creates cross-platform executable
const fakeTscPath = path.join(os.tmpdir(), `fake-tsc-${Date.now()}`)
fs.writeFileSync(fakeTscPath, '#!/usr/bin/env node\nconsole.log("Version 5.3.3");\n...')
fs.chmodSync(fakeTscPath, '755')

// Configure Vitest to use it
const config = `export default defineConfig({
  test: { typecheck: { checker: '${fakeTscPath}' } }
})`

```

Is there a specific tsconfig.json structure or tsc invocation pattern I'm missing that would trigger help text output when called with -p <file>?
Alternatively, is the test executable approach acceptable for this edge case, or would you prefer:


A different testing strategy you have in mind
Modifying how Vitest invokes tsc for this specific test scenario

Ujjwaljain16 commented:
@sheremet-va Thank you for the feedback! I've addressed both points:

## Changes Made

**1.  Static Imports**
Replaced dynamic imports with static imports at the top of the file:
```typescript
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'
import { createFile, runInlineTests } from '../../test-utils'
```
**2. Using createFile Utility**
Replaced manual file creation and cleanup with the createFile() utility:
```
createFile(fakeTscPath, scriptContent)
```
***Test Results***
```
✓ test/typecheck-error.test.ts (1 test) 417ms
  ✓ Typechecker Error Handling (1)
    ✓ throws helpful error when tsc outputs help text (missing config)

Test Files  1 passed (1)
Tests       1 passed (1)
````

sheremet-va commented on packages/vitest/src/typecheck/typechecker.ts:
```diff
@@ -123,6 +123,23 @@ export class Typechecker {
+        + `     test: { typecheck: { tsconfig: 'path/to/tsconfig.json' } }\n`
+        + `  3. Check that the tsconfig file is valid JSON`
+
+      console.error(msg)
```
```suggestion
```

Ujjwaljain16 commented on test/typescript/test/typecheck-error.test.ts:
```diff
@@ -0,0 +1,52 @@
+
+    // Create a temporary directory for our fake tsc
+    const tmpDir = path.join(os.tmpdir(), `vitest-test-${Date.now()}`)
+    fs.mkdirSync(tmpDir, { recursive: true })
```
thanks for review changes made

sheremet-va commented:
Thank you!
